src/WeakValueDict.jl

```diff
@@ -1,9 +1,364 @@
+            # in case "key" already exists in a later collided slot.
+            avail = -index
+         end
+      else
+         x = h.vals[index].value
+         if x === nothing
+            # filled -> missing
+            h.ndirty += 1
+            _deleteindex!(h, index)
+            if avail == 0
+               # ditto
+               avail = -index
+            end
+         elseif key === h.keys[index] || isequal(key, h.keys[index])
+            return index
+         end
+      end
+      index = (index & (sz-1)) + 1
+      iter += 1
+   end
+   avail < 0 && return avail
+   # Check if key is not present, may need to keep searching to find slot
+   while iter <= max(16, sz>>6)
+      if h.slots[index] == 0x1
+         x = h.vals[index].value
+         if x === nothing
+            h.maxprobe = iter
+            return -index
+         end
+      else
+         h.maxprobe = iter
+         return -index
+      end
+      index = (index & (sz-1)) + 1
+      iter += 1
+   end
+   rehash!(h, h.nfilled > 64000 ? sz*2 : sz*4)
+   return ht_keyindex2!(h, key)
+end
+
+### delete! ####
+
+function delete!(h::WeakValueCache, key)
+   index = ht_keyindex(h, key)
+   if index > 0
+      _deleteindex!(h, index)
+   end
+   return h
+end
+
+### getindex / haskey ####
+
+function Base.haskey(h::WeakValueCache, key)
+   index = ht_keyindex(h, key)
+   return index > 0 && h.vals[index].value !== nothing
+end
+
+function Base.getindex(h::WeakValueCache{K, V}, key) where {K, V}
+   index = ht_keyindex(h, key)
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   end
+   throw(KeyError(key))
+end
+
+function Base.get(h::WeakValueCache{K, V}, key, default) where {K, V}
+   index = ht_keyindex(h, key)
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   end
+   return default
+end
+
+### get! ####
+
+function Base.get!(default, h::WeakValueCache{K, V}, key0) where {K, V}
+   key = convert(K, key0)
+   isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+   return Base.get!(default, h, key)
+end
+
+function Base.get!(default, h::WeakValueCache{K, V}, key::K) where {K, V}
+   index = ht_keyindex2!(h, key)
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   else
+      index = -index
+   end
+   age0 = h.age
+   x = convert(V, default())
+   if h.age == age0
+      _setindex!(h, WeakRef(x), key, index)
+   else
+      # the call to default could have changed the internals of h
+      setindex!(h, x, key)
+   end
+   return x
+end
+
+### setindex! ####
+
+function setindex!(h::WeakValueCache{K, V}, v0, key0) where {K, V}
+   key = convert(K, key0)
+   isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+   setindex!(h, v0, key)
+end
+
+# the WeakRef makes key conversion essentially useless
```

Comment:
Do you mean value conversion (not key conversion)? 

I agree that one probably only wants to store the exact object one passed in, and this implicit conversion, will just hide accidental conversions / inefficiencies.

Although strictly speaking the result won't necessarily be useless... but if someone wants this, they might as well insert an *explicit* conversion instead.

Reply:
oops, I added this comment by mistake when I was not thinking clearly. Key conversions are clearly OK in a weak*value*dict/cache.